src/worker/typosquat/cache.rs

```diff
@@ -0,0 +1,100 @@
+use std::{
+    sync::{Arc, Mutex},
+    time::Instant,
+};
+
+use diesel::PgConnection;
+use typomania::{
+    checks::{Bitflips, Omitted, SwappedWords, Typos},
+    Harness,
+};
+
+use crate::swirl::PerformError;
+
+use super::{config, types::TopCrates};
+
+/// Gets the typomania harness for the cached top crates, regenerating it if it is out of date.
+pub(super) fn get_harness(
+    conn: &mut PgConnection,
+) -> Result<Arc<Harness<TopCrates>>, PerformError> {
+    HARNESS_CACHE.get(conn)
+}
+
+static HARNESS_CACHE: Cache = Cache::new();
+
+struct Cache(Mutex<Inner>);
+
+impl Cache {
+    const fn new() -> Self {
+        Self(Mutex::new(Inner::new()))
+    }
+
+    fn get(&self, conn: &mut PgConnection) -> Result<Arc<Harness<TopCrates>>, PerformError> {
+        let mut inner = self.0.lock().unwrap();
+
+        Ok(if let Some(harness) = inner.get() {
+            harness
+        } else {
+            let harness = Arc::new(new_harness(conn)?);
+
+            inner.update(harness.clone());
+            harness
+        })
+    }
+}
+
+#[instrument(skip_all)]
+fn new_harness(conn: &mut PgConnection) -> Result<Harness<TopCrates>, PerformError> {
+    debug!("Rebuilding top crate cache");
+    let start = Instant::now();
+
+    let top_crates = TopCrates::new(conn, config::TOP_CRATES)?;
+
+    // This is essentially the standard set of checks that was implemented by typogard-crates.
+    let harness = Harness::builder()
+        .with_check(Bitflips::new(
+            config::CRATE_NAME_ALPHABET,
+            top_crates.iter_names(),
+        ))
+        .with_check(Omitted::new(config::CRATE_NAME_ALPHABET))
+        .with_check(SwappedWords::new("-_"))
+        .with_check(Typos::new(config::TYPOS.iter().map(|(c, typos)| {
+            (*c, typos.iter().map(|ss| ss.to_string()).collect())
+        })))
+        .build(top_crates);
+
+    let elapsed = Instant::now() - start;
+    debug!(?elapsed, "Top crate cache rebuilt");
+
+    Ok(harness)
+}
+
+struct Inner {
+    harness: Option<Arc<Harness<TopCrates>>>,
+    last_update: Option<Instant>,
+}
+
+impl Inner {
+    const fn new() -> Self {
+        Self {
+            harness: None,
+            last_update: None,
+        }
+    }
+
+    fn get(&self) -> Option<Arc<Harness<TopCrates>>> {
+        if let Some(harness) = &self.harness {
+            if let Some(when) = self.last_update {
+                if when >= Instant::now() - config::CACHE_TTL {
+                    return Some(harness.clone());
+                }
+            }
+        }
+        None
+    }
+
+    fn update(&mut self, harness: Arc<Harness<TopCrates>>) {
+        self.harness = Some(harness);
+        self.last_update = Some(Instant::now());
+    }
```

Comment:
since Heroku restarts our dynos every 24 hours I'm wondering if we even need the TTL/update functionality. if it would simplify the code then we could consider dropping it for now.

Reply:
Yep, good call. I've removed all the cache TTL stuff for now.